### ISSUE TITLE
CSS box-shadow and text-shadow

### ISSUE THREAD

**Issue by SimonSapin** [Kozea/WeasyPrint#13]
WeasyPrint currently does not support the CSS [`box-shadow`](http://www.w3.org/TR/css3-background/#the-box-shadow) or [`text-shadow`](http://www.w3.org/TR/css3-text/#text-shadow) properties.

The main limitation is that these shadow use a Gaussian blur, but cairo (by itself) does not do such blurs. Since the blur can not be expressed as vectors in a PDF file, they need to be rendered to pixel-based images. It would be best to make images at the [surface’s fallback resolution](http://www.cairographics.org/documentation/pycairo/2/reference/surfaces.html#cairo.Surface.get_fallback_resolution). Scipy can do the [gaussian filter](http://docs.scipy.org/doc/scipy/reference/generated/scipy.ndimage.filters.gaussian_filter.html) on a numpy array based on the [image’s data](http://www.cairographics.org/documentation/pycairo/2/reference/surfaces.html#cairo.ImageSurface.get_data).

**Comment by Afoucaul** [comment]
Hi there, I commented on the related PR but I guess it's better suited to post here.

Do you guys still plan on implementing the `box-shadow` property?
I would have loved to see @liZe's implementation from `shadow` merged into `master`, to have at least one implementation upon which one could have improved.
I'm currently trying to merge that branch into `master`, but it's not trivial for me as I don't really know the codebase.

Do you know, as of today, what would be a satisfying implementation for this?
I'm gonna try and make @liZe's implementation usable with today's code.
If this works out well, I'm intending to implement the property in a more efficient and robust way.

**Comment by liZe** [comment]
Hello @Afoucaul.

First of all, thank you for your pull request #859 and for the improvements you want to include in WeasyPrint.

> Do you guys still plan on implementing the `box-shadow` property?

Yes, we do.

> I would have loved to see @liZe's implementation from `shadow` merged into `master`, to have at least one implementation upon which one could have improved.
> I'm currently trying to merge that branch into `master`, but it's not trivial for me as I don't really know the codebase.

Well…

Shadows can be useful, even if they're not great when printed. The problem we have here is that the only solutions we've found are based on raster images, and raster images are bad: they're heavy, they don't scale, they're ugly when printed.

The only possible solution for now would be to have shadows in Cairo. But we don't. And that's sad.

Moreover, PIL is a huge dependency. We're currently trying to have less dependencies (especially the ones with C code), because they're a pain to install for users (see #841). And Azureblur was a proof of concept, it's never been uploaded to PyPI.

Another solution would be to find a replacement for Cairo that handles shadows (see #342 for example).

So… For now, I'll close your pull request, as it relies on both PIL and Azureblur. Even with only PIL, I wouldn't merge it (no offense, I even rejected my own pull request :wink:). I really prefer clean solutions now, as I've had to maintain and deal with a lot of wrong choices before, and it's not funny at all. I know it can be frustrating, but I'm sure that it will force us to find even better solutions.

**Comment by Afoucaul** [comment]
Hello @liZe, thanks for your feedback!

> I'll close your pull request, as it relies on both PIL and Azureblur

Actually this does not need PIL anymore. I kept it in the first place because I integrated your work that used it, but now Azureblur is the only dependency of my implementation.
I might have let some imports but these would be omissions.

> And Azureblur was a proof of concept, it's never been uploaded to PyPI.

But it's straightforward to use, and works out of the box. Seriously, I cloned it, `python setup.py install`ed it, and it just worked. To me, it seems perfect as it is, but do you see any features it lacks, or any fixes it needs?

> I really prefer clean solutions now, as I've had to maintain and deal with a lot of wrong choices before, and it's not funny at all. I know it can be frustrating, but I'm sure that it will force us to find even better solutions.

I strongly agree with your position about keeping things clean. However, reading the state of affairs about shadows in Cairo, it looks like there's never going to be a standard solution, and that everyone out there has resigned themselves to implementing their own way.
Based upon this observation, I believe that the simplest approach is using a well-maintained third-party library, that leads to clean code on the Python side.
That's what I see in Azureblur: the C lib is widely used, and the Python bindings just work and are very intuitive to use.

I will not however try to force a personal decision upon you, the choice is yours.
But if you find out an approach that entirely satisfies you, please let me know, and I'd be glad to help you implement it :)

**Comment by liZe** [comment]
> I might have let some imports but these would be omissions.

Yes, it's only omissions. No problem, I could fix this issue from your PR.

> But it's straightforward to use, and works out of the box. Seriously, I cloned it, `python setup.py install`ed it, and it just worked. To me, it seems perfect as it is, but do you see any features it lacks, or any fixes it needs?

Cloning a repository to install WeasyPrint is OK for you, but it's definitely not a solution for most of our users :smile:.

There are many really big problems with Azureblur:
- It's not even released on PyPI.
- It's not maintained.
- It's not tested.
- It bundles its own (outdated) versions of Moz2D.
- It needs a C / C++ compiler to be built.

> However, reading the state of affairs about shadows in Cairo, it looks like there's never going to be a standard solution, and that everyone out there has resigned themselves to implementing their own way.

Adding blur in a way that's adapted to the many output backends Cairo supports is not a trivial task. As I've said before, generating a raster image for a PDF is a bad solution for me, no matter how good and how simple the code is to do this. I've even rejected my own pull request.

The worst PR I could merge is a raster shadow with no extra libraries (and it really has to be really simple). You **may** be able to do this by adapting [the official blur snippet for Cairo](https://www.cairographics.org/cookbook/blur.c/) into a Python function using CairoCFFI.

**Comment by claell** [comment]
I opened #2849 to implement the lossless zero-blur `box-shadow` subset discussed here and in https://github.com/Kozea/WeasyPrint/pull/149#issuecomment-1254718532.

It supports multiple outer and inset shadows, offsets, spread, colors and `currentColor`, rounded corners, transparent-box knockout, and the relevant collapsed-table behavior. The output remains vector-only; nonzero blur radii are rejected with a clear warning. `text-shadow`, `filter: drop-shadow()`, and raster blur policy remain separate follow-up concerns.

The full rationale, implementation notes, generic test coverage, and verification results are in the pull request.

**Comment by claell** [comment]
Following up on https://github.com/Kozea/WeasyPrint/pull/2846#issuecomment-5011728769:

My current thoughts are the following:

* PDFs seem to have no support for this, thus a lossless implementation is kind of impossible.
* When there is no blur, this can actually be done in a lossless way, thus is preferable over a rasterization approach IMHO, thus #2849.
* For all other shadows with a blur radius, we need a different path:
  * Important: Have this opt-in. By default, such shadows are ignored, keeping default rendering as is. Only if a user wants to explicitly enable support for this, render it. Potentially show a hint to users how to activate rendering in case they have such shadows in their input.
  * Show warnings for problematic ones, in general have warnings about potential problems in case a user activates that option.
  * Implement good rasteritation defaults with support for custom settings if a user wants to have.


> The worst PR I could merge is a raster shadow with no extra libraries (and it really has to be really simple).

This seems to be potentially outdated, now that Pillow has become a project dependency, right?

In general, from what I saw, one major concern that might block this is maintainability? That probably needs discussion (what is okay to add, what only creates debt, structuring, ...).
